examples/deal.II/bp.h

```diff
@@ -0,0 +1,1165 @@
+              qdata[i + Q * 5] * ug0 + qdata[i + Q * 1] * ug1 + qdata[i + Q * 3] * ug2;
+            vg[i + Q * 2 + Q * 3 * 2] =
+              qdata[i + Q * 4] * ug0 + qdata[i + Q * 3] * ug1 + qdata[i + Q * 2] * ug2;
+          }
+        }
+        break;
+    }
+  return 0;
+}
+
+
+
+/**
+ * Base class of operators.
+ */
+template <typename Number>
+class OperatorBase
+{
+public:
+  /**
+   * deal.II vector type
+   */
+  using VectorType = LinearAlgebra::distributed::Vector<Number>;
+
+  /**
+   * Initialize vector.
+   */
+  virtual void
+  reinit() = 0;
+
+  /**
+   * Perform matrix-vecor product
+   */
+  virtual void
+  vmult(VectorType &dst, const VectorType &src) const = 0;
+
+  /**
+   * Initialize vector.
+   */
+  virtual void
+  initialize_dof_vector(VectorType &vec) const = 0;
+
+  /**
+   * Compute inverse of diagonal.
+   */
+  virtual void
+  compute_inverse_diagonal(VectorType &diagonal) const = 0;
+};
+
+
+/**
+ * Operator implementation using libCEED.
+ */
+template <int dim, typename Number>
+class OperatorCeed : public OperatorBase<Number>
+{
+public:
+  using VectorType = typename OperatorBase<Number>::VectorType;
+
+  /**
+   * Constructor.
+   */
+  OperatorCeed(const Mapping<dim>              &mapping,
+               const DoFHandler<dim>           &dof_handler,
+               const AffineConstraints<Number> &constraints,
+               const Quadrature<dim>           &quadrature,
+               const BPType                    &bp)
+    : mapping(mapping)
+    , dof_handler(dof_handler)
+    , constraints(constraints)
+    , quadrature(quadrature)
+    , bp(bp)
+  {
+    reinit();
+  }
+
+  /**
+   * Destructor.
+   */
+  ~OperatorCeed()
+  {
+    CeedOperatorDestroy(&op_apply);
+    CeedQFunctionDestroy(&qf_apply);
+    CeedQFunctionContextDestroy(&build_ctx);
+    CeedVectorDestroy(&q_data);
+    CeedElemRestrictionDestroy(&q_data_restriction);
+    CeedElemRestrictionDestroy(&sol_restriction);
+    CeedBasisDestroy(&sol_basis);
+    CeedDestroy(&ceed);
+  }
+
+  /**
+   * Initialized internal data structures, particularly, CEED.
+   */
+  void
+  reinit() override
+  {
+    const auto &tria = dof_handler.get_triangulation();
+    const auto &fe   = dof_handler.get_fe();
+
+    const auto n_components = fe.n_components();
+
+    if (bp == BPType::BP1 || bp == BPType::BP3 || bp == BPType::BP5)
+      {
+        AssertThrow(n_components == 1, ExcInternalError());
+      }
+    else
+      {
+        AssertThrow(n_components == dim, ExcInternalError());
+      }
+
+    // 1) create CEED instance -> "MatrixFree"
+    const char *ceed_spec = "/cpu/self/avx/blocked";
+    CeedInit(ceed_spec, &ceed);
+
+    // 2) create shape functions -> "ShapeInfo"
+    const unsigned int fe_degree  = fe.tensor_degree();
+    const unsigned int n_q_points = quadrature.get_tensor_basis()[0].size();
+    CeedBasisCreateTensorH1Lagrange(
+      ceed, dim, n_components, fe_degree + 1, n_q_points, CEED_GAUSS, &sol_basis);
```

Comment:
> It would be cool if we could build the CeedBasis from information from the deal.II FE space

I like this suggestion. I didn't feel too happy with the approach I have taken here.

Reply:
What do you think about us focusing on some small tweaks to get this new code running in our CI pipeline in this PR and then making this update in a follow-up? I don't think it should be too much extra work, but I don't want to block merging with a bunch of little changes